core/src/main/scala/higherkindness/mu/rpc/srcgen/avro/AvroSrcGenerator.scala

```diff
@@ -1,198 +1,90 @@
+        inputFile: File,
+        serializationType: Model.SerializationType
+    ): ErrorsOr[Generator.Output] = {
+      val nativeAvroProtocol: ErrorsOr[Protocol] =
+        (new FileInputParser)
+          .getSchemaOrProtocols(inputFile, Standard, classStore, classLoader)
+          //multiple protocols are returned when imports are present.
+          //We assume the first one is the one defined in our file
+          .collectFirst { case Right(protocol) =>
+            protocol
+          }
+          .toValidNel(s"No protocol definition found in ${inputFile}")
+
+      val skeuomorphAvroProtocol: ErrorsOr[AvroProtocol[Mu[AvroF]]] =
+        nativeAvroProtocol.andThen(p =>
+          Try(AvroProtocol.fromProto[Mu[AvroF]](p)).toEither.toValidatedNel
+            .leftMap { ts: NonEmptyList[Throwable] =>
+              ts.map(_.getMessage)
+            }
+        )
+
+      val source = skeuomorphAvroProtocol.andThen { sap =>
+        val muProtocol: MuProtocol[Mu[MuF]] =
+          MuProtocol.fromAvroProtocol(compressionType, useIdiomaticEndpoints)(sap)
+
+        val outputFilePath = getPath(sap)
+
+        val streamCtor: (Type, Type) => Type.Apply = streamingImplementation match {
+          case Fs2Stream => { case (f, a) =>
+            t"_root_.fs2.Stream[$f, $a]"
+          }
+          case MonixObservable => { case (_, a) =>
+            t"_root_.monix.reactive.Observable[$a]"
+          }
         }
-      packageLines ++ importLines ++ messageLines ++ serviceLines
-    }
 
-    outputPath -> outputCode
-  }
+        val stringified: ErrorsOr[List[String]] =
+          codegen
+            .protocol(muProtocol, streamCtor)
+            .toValidatedNel
+            .map(_.syntax.split("\n").toList)
 
-  private def validateRequest(request: Schema): ErrorsOr[String] = {
-    val requestArgs = request.getFields.asScala
-    requestArgs.toList match {
-      case Nil =>
-        s"$DefaultRequestParamName: $EmptyType".validNel
-      case x :: Nil if x.schema.getType == Schema.Type.RECORD =>
-        s"${requestArgs.head.name}: ${requestArgs.head.schema.getFullName}".validNel
-      case _ :: Nil =>
-        s"RPC method request parameter '${requestArgs.head.name}' has non-record request type '${requestArgs.head.schema.getType}'".invalidNel
-      case _ =>
-        s"RPC method ${request.getName} has more than 1 request parameter".invalidNel
+        stringified.map(Generator.Output(outputFilePath, _))
+      }
+      source
     }
   }
-
-  private def validateResponse(response: Schema): ErrorsOr[String] = {
-    response.getType match {
-      case Schema.Type.NULL =>
-        EmptyType.validNel
-      case Schema.Type.RECORD =>
-        s"${response.getNamespace}.${response.getName}".validNel
-      case _ =>
-        s"RPC method response parameter has non-record response type '${response.getType}'".invalidNel
-    }
+  private def getPath(p: AvroProtocol[Mu[AvroF]]): Path = {
+    val fst :: rest =
```

Comment:
maybe I'm obtuse, but it's not immediately clear what the `fst` variable name represents here (or `rest` for that matter...).  Is it supposed to represent like a [finite-state transducer](https://en.wikipedia.org/wiki/Finite-state_transducer)?  Or something else that I don't totally grok?  Either way, I'd love an explanation just for my understanding :)

Reply:
Nothing like that,  I think my mind was more thinking "first part of path" and "rest of path"...a more idiomatic name would be `val head :: tail`. The only reason i'm even splitting this up is the `java.nio.file.Paths.get` type signature (I switched paths from `String` to `Path` where I saw them).

I'll change it to be more clear

Reply:
looks great!